File: lc_soln/0146_lru_cache/solution.py
class Node:
    def __init__(self, key: int = -1, val: int = -1):
        self.key = key
        self.val = val
        self.prev = None
        self.next = None


def remove_node(node: Node) -> Node:
    node.prev.next = node.next
    node.next.prev = node.prev
    node.prev = None
    node.next = None
    return node


def insert_after(node: Node, target: Node) -> Node:
    node.prev = target
    node.next = target.next
    target.next.prev = node
    target.next = node
    return node


class LRUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.size = 0
        self.head = Node()
        self.tail = Node()
        self.head.next = self.tail
        self.tail.prev = self.head

        self.key_to_node = {}

    def get(self, key: int) -> int:
        node = self.key_to_node.get(key)
        if node is None:
            return -1
        node = remove_node(node)
        node = insert_after(node, self.head)
        return node.val

    def put(self, key: int, value: int) -> None:
        node = self.key_to_node.get(key)
        if node is None:
            if self.size == self.capacity:
                removed = remove_node(self.tail.prev)
                del self.key_to_node[removed.key]
                self.size -= 1

            node = Node(key=key, val=value)
            node = insert_after(node, self.head)
            self.key_to_node[key] = node
            self.size += 1
        else:
            node.val = value
            node = remove_node(node)
            node = insert_after(node, self.head)


def main(): ...


if __name__ == "__main__":
    main()

File: lc_soln/0146_lru_cache/solution.cs
public class LRUCacheNode(int key = -1, int val = -1)
{
    public int Key { get; set; } = key;
    public int Val { get; set; } = val;
    public LRUCacheNode Prev { get; set; } = null;
    public LRUCacheNode Next { get; set; } = null;
}

public class LRUCache
{
    private int _size = 0;
    private readonly Dictionary<int, LRUCacheNode> _keyToNode = [];
    private readonly int _capacity;
    private readonly LRUCacheNode _head;
    private readonly LRUCacheNode _tail;

    public LRUCache(int capacity)
    {
        _capacity = capacity;
        _head = new LRUCacheNode();
        _tail = new LRUCacheNode();
        _head.Next = _tail;
        _tail.Prev = _head;
    }

    static LRUCacheNode Remove(LRUCacheNode node)
    {
        node.Prev.Next = node.Next;
        node.Next.Prev = node.Prev;
        node.Prev = null;
        node.Next = null;
        return node;
    }

    static LRUCacheNode InsertAfter(LRUCacheNode node, LRUCacheNode target)
    {
        node.Prev = target;
        node.Next = target.Next;
        target.Next.Prev = node;
        target.Next = node;
        return node;
    }

    public int Get(int key)
    {
        if (_keyToNode.TryGetValue(key, out LRUCacheNode node))
        {
            node = Remove(node);
            node = InsertAfter(node, _head);
            return node.Val;
        }
        return -1;
    }

    public void Put(int key, int value)
    {
        if (_keyToNode.TryGetValue(key, out LRUCacheNode node))
        {
            node.Val = value;
            node = Remove(node);
            node = InsertAfter(node, _head);
        }
        else
        {
            if (_size == _capacity)
            {
                LRUCacheNode removed = Remove(_tail.Prev);
                _keyToNode.Remove(removed.Key);
                _size--;
            }
            node = new LRUCacheNode(key, value);
            node = InsertAfter(node, _head);
            _keyToNode[key] = node;
            _size++;
        }
    }
}
